### DISCUSSION TITLE
Question regarding "SQL (Relational) Databases"

### DISCUSSION THREAD

**Discussion by gerazenobi** [fastapi/fastapi#6721] [Questions] [answered]
Hi there, 

First of all just wanted to say great job providing such a rich and clear documentation 👍 

I had a question regarding the section ["SQL (Relational) Databases"](https://fastapi.tiangolo.com/tutorial/sql-databases/?h=sq): I am correct to **(A)** assume all the examples in this section are blocking IO operations? It threw me off a bit considering we are in the context of a cooperative asynchronous framework and wanted to confirm I was not missing/miss interpreting anything. 
**(B)** Furthermore it would seem all the db code generated by the [project generator](https://github.com/tiangolo/full-stack-fastapi-postgresql) is also blocking rather than async ? Would that mean this generated code is not production ready and needs to be adjusted to be async?

Is there a chance the documentation reflects SQLAlchemy Core/ORM Asyncio [support](https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html) soon rather than having to use another library's API as proposed in [Async SQL (Relational) Databases](https://fastapi.tiangolo.com/advanced/async-sql-databases/?h=sq) ?

Personally, I find a bit confusing the documentation has 2 sections for RDMS databases integration:  _blocking_ VS _non blocking_ considering we should never be using the first one unless I am missing a use case?

Thanks in advance for your time.

**Comment by ArcLightSlavik** [comment, marked as answer]
A) Yes all of those are 'blocking IO operations'
B) Yes all of the code there is also blocking, but it is production ready.

PR for the tutorial with async sqlalchemy: <https://github.com/tiangolo/fastapi/pull/2331>

SQLAlchemy is by far the most popular ORM and it didn't have async support until october of 2020. Not having that tutorial in would impose a lot of questions of the type: "how do you integrate fastapi with sqlalchemy". Furthermore just because it's not async doesn't mean u can't use it, FastAPI supports non async functions just as well as async. Allowing people to integrate in steps, e.g first change to FastAPI and then change the actual logic of requests.

**Comment by gerazenobi** [comment]
Thanks for your answer @ArcLightSlavik 👍 

After reading other [questions & answers](https://github.com/tiangolo/fastapi/issues/260#issuecomment-495945630) and digging a bit deeper in the FastAPI [documentation](https://fastapi.tiangolo.com/async/#very-technical-details) I understood finally how was that the framework was compatible with blocking code/libraries.
I guess it was just the order in the documentation that threw me off.

Gonna take a look at your PR !

Cheers.

**Comment by tiangolo** [comment]
Thanks for the help here @ArcLightSlavik ! 👏  🙇 

Thanks for reporting back and closing the issue @gerazenobi  👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
